Makefile

```diff
@@ -154,3 +182,26 @@ check-shell-scripts:
 clean:
 	$(RM) $(BUILDFILE_SERVER) $(BUILDFILE_NIGHTLY_SERVER) $(BUILDFILE_AD_SERVER) $(BUILDFILE_NIGHTLY_AD_SERVER) $(BUILDFILE_CLIENT) $(BUILDFILE_TOOLBOX)
 .PHONY: clean
+
+# _img_build is an "internal" rule to make the building of samba-container
+# images regular and more "self documenting". A makefile.foo that includes
+# this Makefile can add build rules using _img_build as a building block.
+#
+# The following arguments are expected to be supplied when "calling" this rule:
+# BUILD_ARGS: the default build arguments
+# EXTRA_BUILD_ARGS: build args supplied by the user at "runtime"
+# SHORT_NAME: a local name for the image
+# REPO_NAME: a global name for the image
+# SRC_FILE: path to the Containerfile (Dockerfile)
+# DIR: path to the directory holding image contents
+# BUILDFILE: path to a temporary file tracking build state
+_img_build:
+	$(BUILD_CMD) \
+		$(BUILD_ARGS) \
+		$(EXTRA_BUILD_ARGS) \
+		--tag $(SHORT_NAME) \
+		--tag $(REPO_NAME) \
+		-f $(SRC_FILE) \
+		$(DIR)
+	$(CONTAINER_CMD) inspect -f '{{.Id}}' $(SHORT_NAME) > $(BUILDFILE)
+.PHONY: _img_build
```

Comment:
This is a valid solution. Personally, I prefer using [call-function](https://www.gnu.org/software/make/manual/html_node/Call-Function.html) instead of invoking sub-make rule. Something like:

```
define _img_build2
	$(BUILD_CMD) $(1) $(2) --tag $(3) --tag $(4) -f $(5) $(6)
	$(CONTAINER_CMD) inspect -f '{{.Id}}' $(3) > $(7)
endef
```

And use it with:

```
build-toolbox: $(BUILDFILE_TOOLBOX)
.PHONY: build-toolbox
$(BUILDFILE_TOOLBOX): Makefile $(TOOLBOX_SRC_FILE)
	$(call _img_build2, , $(EXTRA_BUILD_ARGS), \
		$(TOOLBOX_NAME), $(TOOLBOX_REPO_NAME), \
		$(TOOLBOX_SRC_FILE), $(TOOLBOX_DIR), $(BUILDFILE_TOOLBOX) )
```

I assume its a matter of personal preference. Both solutions are valid.

Reply:
I went looking for how to do that but the gnu make docs are a bit meh in my opinion and I wasn't fully up on functions in make. The other issue I have with call-function is that I much prefer named to positional parameters and since I didn't see a way to do named params with a function I reverted to the approach I knew best.